saq/job.py

```diff
@@ -258,3 +268,77 @@ def replace(self, job):
         """Replace current attributes with job attributes."""
         for field in job.__dataclass_fields__:
             setattr(self, field, getattr(job, field))
+
+
+class PeriodicHeartbeat:
+    """Class to manage agent Heartbeats while Job is running"""
+
+    def __init__(self, job: Job) -> None:
+        self._running: bool = True
+        self._heartbeat_task: typing.Optional[asyncio.Task] = None
+        self.job = job
+        self.heartbeat_enabled: bool = job.heartbeat and job.heartbeat > 0
+        self.time_between_ticks = (
+            max(round(job.heartbeat / 2), 1) if self.heartbeat_enabled > 0 else 0
+        )
+
+    async def start(self, func: typing.Awaitable):
+        """start
+
+        Executes function with a periodic heartbeat.
+        Args:
+            func (typing.Awaitable): _description_
+
+        Returns:
+            _type_: _description_
+        """
+        if self.heartbeat_enabled:
+            logger.info(
+                "starting heartbeat service for %s. Ticking every %s second(s)",
+                self.job.id,
+                self.time_between_ticks,
+            )
+        self._heartbeat_task = asyncio.create_task(self._periodically_publish())
+        executed_func: typing.Optional[typing.Any] = None
+        try:
+            executed_func = await func
+            self._heartbeat_task.cancel()
+            await self._heartbeat_task
+        except asyncio.CancelledError:
+            pass
```

Comment:
This will catch _all_ CancelledErrors, even those that might've been raised elsewhere (for example, if the worker is stopped: https://github.com/tobymao/saq/blob/master/saq/worker.py#L127).

I think that could potentially silence exceptions and make the worker think the job completed successfully. No bueno.